### DISCUSSION TITLE
How do I filter the  data and in sever

### DISCUSSION THREAD

**Discussion by MrShakila** [serverpod/serverpod#1330] [Q&A]
I have custom function that will get the user location and if the location is near 5 km then  I need to send the filtered dog list .
currently I tried using where but Its not working

```
Future<List<AdoptPet>> getNearDog(Session session, LatLang userLatLang) async {
    return AdoptPet.find(
      session,
      limit: 10,
      where: (adoptPet) {
        double thresholdDistance = 5;
        if (isNear(
            userLatLang.latitude,
            userLatLang.longitude,
            double.parse(adoptPet.latitude.toString()),
            double.parse(adoptPet.longitude.toString()),
            thresholdDistance)) {
         return adoptPet.createdAt; // this will return all the data not filters
        }
      
      },
    );
  }
}

How can I do that ?
```

**Comment by SandPod** [comment]
Hey @MrShakila 👋 

The `where` parameter is an `AdoptPetExpresionBuilder` that has as its only parameter the `AdoptPetTable` and returns an `Expression`. It is used to build an expressions used to filter what rows are fetched from the database.

If you look at the `toString()` methods you call they only return the name of the column used in the database and not the value of that column.

Unfortunately Serverpods database operations doesn't have built int support for measuring euclidian distance between two points at this time.

One possible solution could be to implement a crude filter for what rows you would like to fetch from the database filtering on the Chebyshev distance:
``` dart
    return AdoptPet.find(
      session,
      limit: 10,
      where: (adoptPet) {
        double thresholdDistance = 5;
        return adoptPet.latitude <= (userLatLang.latitude + thresholdDistance) &
            adoptPet.latitude >= (userLatLang.latitude - thresholdDistance) &
            adoptPet.longitude <= (userLatLang.longitude + thresholdDistance) &
            adoptPet.longitude >= (userLatLang.longitude - thresholdDistance);
        }
      },
    );
```
    
And then further filter the result if you would like to measure the euclidian distance on the entities returned.

(We recently added support for `.inBetween(...)`, PR: https://github.com/serverpod/serverpod/commit/a4c2fcfdc5a19ab52702ef96d0364eb75b165591, operations that could be helpful here, but it has not been included in an official release yet.)

Another solution could be to implement your own [raw sql query](https://docs.serverpod.dev/concepts/database-communication#executing-raw-queries) and do the desired operations there. Postgres has support for a wide range of operations.
